keymapper/injection/global_uinputs.py

```diff
@@ -0,0 +1,139 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with key-mapper.  If not, see <https://www.gnu.org/licenses/>.
+
+
+import evdev
+import keymapper.utils
+import keymapper.exceptions
+from keymapper.logger import logger
+
+DEV_NAME = "key-mapper"
+DEFAULT_UINPUTS = {
+    # for event codes see linux/input-event-codes.h
+    "keyboard": {
+        evdev.ecodes.EV_KEY: list(evdev.ecodes.KEY.keys() & evdev.ecodes.keys.keys())
+    },
+    "gamepad": {
+        evdev.ecodes.EV_KEY: [*range(0x130, 0x13f)],  # BTN_SOUTH - BTN_THUMBR
+        evdev.ecodes.EV_ABS: [*range(0x00, 0x06), *range(0x10, 0x12)]  # 6-axis and 1 hat switch
+    },
+    "mouse": {
+        evdev.ecodes.EV_KEY: [*range(0x110, 0x118)],  # BTN_LEFT - BTN_TASK
+        evdev.ecodes.EV_REL: [*range(0x00, 0x0a)]  # all REL axis
+    }
+}
+
+
+class UInput(evdev.UInput):
+    def __init__(self, *args, **kwargs):
+        logger.debug(f"creating UInput device: '{kwargs['name']}'")
+        super().__init__(*args, **kwargs)
+
+    def can_emit(self, event):
+        """check if an event can be emitted by the uinput
+
+        Wrong events might be injected if the group mappings are wrong
+        """
+        # TODO check for event value especially for EV_ABS
+        return event[1] in self.capabilities().get(event[0], [])
+
+
+class FrontendUInput:
+    """Uinput which can not actually send events, for use in the frontend"""
+    def __init__(self, *args, **kwargs):
+        defaults = {  # see https://python-evdev.readthedocs.io/en/latest/apidoc.html#module-evdev.uinput
+            "events": None,
+            "name": 'py-evdev-uinput',
+            #"vendor": 1,
+            #"product": 1,
+            #"version": 1,
+            #"bustype": 3,
+            #"devnode": '/dev/uinput',
+            #"phys": 'py-evdev-uinput',
+            }
+        logger.debug(f"creating fake UInput device: '{kwargs['name']}'")
+        for key, value in defaults.items():
+            try:
+                setattr(self, key, kwargs[key])
+            except KeyError:
+                setattr(self, key, value)
+    
+    def capabilities(self):
+        return self.events
+
+
+class GlobalUInputs:
+    """Manages all uinputs that are shared between all injection processes."""
+
+    def __init__(self):
+        self.devices = {}
+
+        if keymapper.utils.is_service():
+            self._uinput_factory = UInput
+        else:
+            self._uinput_factory = FrontendUInput
+
+    def prepare(self, force_service=False):
+        """Generate uinputs.
+
+        This has to be done in the main process before injections start.
+        """
+
+        # TODO: remove force_service we should find a way to patch keymapper.utils.is_service() in the tests
```

Comment:
@sezanzeb is there a good way to archive this? we need the `force_service` only in the tests because `is_service()` always returns `False`

Reply:
and there are some more problems where I would like to hear your thoughts: 

- the fake `UInput` class in `tests/test.py` does not work with `global_uinputs`: when running `test_injector.TestInjector.test_injector` it fails at line 641 with `5 != 11` even though the debug log shows all 11 events. I think all the forwarded events end up in the `uinput_write_history_pipe` but not the events mapped to a global UInput 
- `test_injector.TestInjector.test_store_permutations_for_macros` stops after a `ERROR consumer_control.py:118: The async_read_loop for "/dev/input/event10" stopped early` and never continues. I have no clue where I would start to debug this

Reply:
yes. Something like

```py
with patch.object(sys, "argv", ["key-mapper-service"]):
    ...
```

can probably be used in tests to make is_service return true

---

Is `patch_evdev` happening before `global_uinputs` creates its UInputs? 

---

The async_read_loop error might happen because `async_read_loop` of `InputDevice` in `test.py` finishes reading all events that are configured in the test. It has to stop in order for tests to continue. During runtime it would be an error, in tests it is expected.

Reply:
Put a few seconds of extra sleep before `history = read_write_history_pipe()` and see what happens. Maybe key-mapper is not handling the events fast enough and there is some bottleneck somewhere

Reply:
An alternative to `patch_evdev` with `evdev.UInput = UInput` might be to write those patches into the original `UInput` class:

```
evdev.UInput.write = uinput_write_patch
...
```

This might be more robust, but if that really works the way I think it works remains to be validated.

But if this is not the root cause then there is no priority here.

Reply:
> Is `patch_evdev` happening before `global_uinputs` creates its UInputs?

yes, I run `global_uinputs.prepare()` in the `cleanup()` function and the debug output contains a `grey_log(f"{(type, code, value)} written")` (line 435, test.py) line for each event

> I think you need to first use `push_events` with all the events you intend to read in your test, and then run the injector.

I did not touch the `test_store_permutations_for_macros` other than fixing a call to `mapping.change()` and the test does not read or write any events.

Reply:
I'll check it out in detail tomorrow evening

Reply:
> I'll check test_store_permutations_for_macros out in detail tomorrow evening

no need to, I found the error `injector._construct_capabilities()` is not used but the test relies on it to stop the injection before it starts.

Reply:
I wonder if there is anything we can do to make writing tests easier. Everything seems really complicated

Reply:
yeah, it is quiet tedious to work on these tests.
> Put a few seconds of extra sleep before `history = read_write_history_pipe()` in `test_injector` and see what happens.

no difference at all.
I updated `write()` in the fake test UInput to print the pipe object it is writing to:

```
37622 22:46:41.522199 DEBUG consumer_control.py:83: Starting to listen for events from /dev/input/event20, fd 44
async_read_loop "Bar Device" "/dev/input/event20" InputEvent(1, 8, 1)
37622 22:46:41.522659 SPAM keycode_mapper.py:539: ((1, 8, 1)) ------------------- forwarding
(1, 8, 1) written to <multiprocessing.connection.Connection object at 0x7fb3f3d0cbb0>
async_read_loop "Bar Device" "/dev/input/event20" InputEvent(1, 9, 1)
37622 22:46:41.533072 SPAM keycode_mapper.py:516: ((1, 8, 1), (1, 9, 1)) -------- maps to macro (k(KEY_Q).k(w), keyboard)
37622 22:46:41.533208 DEBUG keycode_mapper.py:329: macro_sending (1, 101, 1)
(1, 101, 1) written to <multiprocessing.connection.Connection object at 0x7fb3f5b84fd0>
async_read_loop "Bar Device" "/dev/input/event20" InputEvent(1, 8, 0)
37622 22:46:41.543643 DEBUG keycode_mapper.py:329: macro_sending (1, 101, 0)
(1, 101, 0) written to <multiprocessing.connection.Connection object at 0x7fb3f5b84fd0>
37622 22:46:41.543816 SPAM keycode_mapper.py:463: ((1, 8, 0)) ------------------- forwarding release
(1, 8, 0) written to <multiprocessing.connection.Connection object at 0x7fb3f3d0cbb0>
37622 22:46:41.554053 DEBUG keycode_mapper.py:329: macro_sending (1, 102, 1)
(1, 102, 1) written to <multiprocessing.connection.Connection object at 0x7fb3f5b84fd0>
async_read_loop "Bar Device" "/dev/input/event20" InputEvent(1, 9, 0)
37622 22:46:41.554309 SPAM keycode_mapper.py:435: ((1, 9, 0)) ------------------- releasing macro
37622 22:46:41.554362 SPAM keycode_mapper.py:448: ((1, 9, 0)) ------------------- releasing key
37622 22:46:41.564650 DEBUG keycode_mapper.py:329: macro_sending (1, 102, 0)
(1, 102, 0) written to <multiprocessing.connection.Connection object at 0x7fb3f5b84fd0>
async_read_loop "Bar Device" "/dev/input/event20" InputEvent(3, 16, -1)
37622 22:46:41.564925 SPAM keycode_mapper.py:531: ((3, 16, -1)) ----------------- maps to (100, keyboard)
(1, 100, 1) written to <multiprocessing.connection.Connection object at 0x7fb3f5b84fd0>
async_read_loop "Bar Device" "/dev/input/event20" InputEvent(3, 16, 0)
37622 22:46:41.575298 SPAM keycode_mapper.py:454: ((3, 16, 0)) ------------------ releasing (100, keyboard)
(1, 100, 0) written to <multiprocessing.connection.Connection object at 0x7fb3f5b84fd0>
async_read_loop "Bar Device" "/dev/input/event20" InputEvent(1, 10, 1)
37622 22:46:41.585662 SPAM keycode_mapper.py:539: ((1, 10, 1)) ------------------ forwarding
(1, 10, 1) written to <multiprocessing.connection.Connection object at 0x7fb3f3d0cbb0>
async_read_loop "Bar Device" "/dev/input/event20" InputEvent(1, 10, 0)
37622 22:46:41.596140 SPAM keycode_mapper.py:463: ((1, 10, 0)) ------------------ forwarding release
(1, 10, 0) written to <multiprocessing.connection.Connection object at 0x7fb3f3d0cbb0>
async_read_loop "Bar Device" "/dev/input/event20" InputEvent(3124, 3564, 6542)
(3124, 3564, 6542) written to <multiprocessing.connection.Connection object at 0x7fb3f3d0cbb0>
37622 22:46:41.616714 ERROR consumer_control.py:118: The async_read_loop for "/dev/input/event20" stopped early
```

It looks like there are in fact two different pipes (at `0x7fb3f3d0cbb0` and `0x7fb3f5b84fd0`) I think this is currently my main issue, as it will cause a lot of tests to fail

Reply:
since `uinput_write_history` is a global variable it probably means there are two processes running. Hence the different identifiers.

It seems that the `b0` pipe is used for forwarding, and the `d0` pipe by the `keycode_mapper`

Reply:
You are using `global_uinputs.prepare` in `cleanup`, but `TestInjector` is using `quick_cleanup`. Could this be a contributing factor?

Reply:
If you have ideas for improving the tests in general, please throw them into https://github.com/sezanzeb/key-mapper/issues/246

Reply:
> You are using global_uinputs.prepare in cleanup, but TestInjector is using quick_cleanup. Could this be a contributing factor?

maybe but I think this is a bug in the overall design. Even in the main branch there are two pipes created but since all uinputs get created in the injector this is not a problem on main.

Using the traceback module we can see that the first pipe is created at the very beginning, this was previously never used. The second one gets created by the `import` statement in `test_injector.py`:
 ```
$ python tests/test.py test_injector.TestInjector.test_injector
creating pipe (<multiprocessing.connection.Connection object at 0x7f36ee8c5520>, <multiprocessing.connection.Connection object at 0x7f36ef3d9070>)
File "/home/jonas/dev/key-mapper/tests/test.py", line 111, in <module>
    for line in traceback.format_stack():
41672 00:12:12.313981 SPAM pipe.py:62: Creating new pipe for "/tmp/key-mapper-jonas/results"
41672 00:12:12.314752 SPAM pipe.py:62: Creating new pipe for "/tmp/key-mapper-jonas/commands"
cleanup
.
.
.
41672 00:12:12.401945 INFO groups.py:448: Found "Foo Device", "Foo Device 2", "Bar Device", "gamepad", "key-mapper Bar Device"
41672 00:12:12.402037 DEBUG global_uinputs.py:48: creating UInput device: 'key-mapper keyboard'
41672 00:12:12.402095 DEBUG global_uinputs.py:48: creating UInput device: 'key-mapper gamepad'
41672 00:12:12.402127 DEBUG global_uinputs.py:48: creating UInput device: 'key-mapper mouse'
creating pipe (<multiprocessing.connection.Connection object at 0x7f36ed5c0310>, <multiprocessing.connection.Connection object at 0x7f36ed5c03a0>)
File "/home/jonas/dev/key-mapper/tests/test.py", line 687, in <module>
    main()
File "/home/jonas/dev/key-mapper/tests/test.py", line 667, in main
    testsuite = unittest.defaultTestLoader.loadTestsFromNames(
File "/usr/lib/python3.9/unittest/loader.py", line 220, in loadTestsFromNames
    suites = [self.loadTestsFromName(name, module) for name in names]
File "/usr/lib/python3.9/unittest/loader.py", line 220, in <listcomp>
    suites = [self.loadTestsFromName(name, module) for name in names]
File "/usr/lib/python3.9/unittest/loader.py", line 154, in loadTestsFromName
    module = __import__(module_name)
File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
File "<frozen importlib._bootstrap>", line 972, in _find_and_load_unlocked
File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
File "<frozen importlib._bootstrap>", line 972, in _find_and_load_unlocked
File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
File "<frozen importlib._bootstrap_external>", line 850, in exec_module
File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
File "/home/jonas/dev/key-mapper/tests/testcases/test_injector.py", line 70, in <module>
    from tests.test import (
File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
File "<frozen importlib._bootstrap_external>", line 850, in exec_module
File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
File "/home/jonas/dev/key-mapper/tests/test.py", line 111, in <module>
    for line in traceback.format_stack():
quick cleanup
41672 00:12:12.410838 DEBUG reader.py:196: Sending close msg to helper
```
I would guess this is caused by tests/test.py being run as `__main__` and then later imported as a module. The entry in https://docs.python.org/3/library/sys.html#sys.modules is only created once tests is loaded as a module, therefore we end up with two pipes

Reply:
okay, turns out my guess was correct. This is an important lessons learned. I was able to fix it by moving the `if name == main` to the top and importing the file to itself:

```
import os
import sys

# make sure the "tests" module visible
sys.path.append(os.getcwd())
if __name__ == "__main__":
    # import this file to itself to make sure is not run twice and all global variables end up in sys.modules
    import tests.test
    tests.test.main()
    sys.exit(0)
```

Reply:
and of course we are not alone with this problem:
https://stackoverflow.com/questions/13181559/importing-modules-main-vs-import-as-module

Reply:
Nice! Please add the link to the stackoverflow post as a comment to the fix

We also have a circular import of some sort. The discoverer in test.py runs test files that import test.py again.

Reply:
I think that is totally intended every test has to import `tests.test` in order to get access `quick_cleanup` and other stuff. But it is not circular until we import a `tests.testcases. ...` back into `tests.test` which we don't.

Reply:
Ok. I was thinking, maybe the test utility functions like `quick_cleanup` should be moved into a separate file.

But if it doesn't break anything the way it is now I guess it's allright